# **Ejercicios de pair programming Módulo 2 Sprint 1**
## **Estadística: Lección 1**
### **Introducción**

---

In [8]:
import numpy as np
import pandas as pd

1. Buscar un conjunto de datos a analizar:
 - Se recomienda que el conjunto de datos a analizar tenga variables numéricas y categóricas. Ya que también se analizarán.


In [30]:
games = pd.read_csv("data/games.csv", index_col=0)
games.head()

Title  Release Date  \
0                               Elden Ring  Feb 25, 2022   
1                                    Hades  Dec 10, 2019   
2  The Legend of Zelda: Breath of the Wild  Mar 03, 2017   
3                                Undertale  Sep 15, 2015   
4                            Hollow Knight  Feb 24, 2017   

                                                Team  Rating  Times Listed  \
0     ['Bandai Namco Entertainment', 'FromSoftware']     4.5         39000   
1                               ['Supergiant Games']     4.3         29000   
2  ['Nintendo', 'Nintendo EPD Production Group No...     4.4         43000   
3                                 ['tobyfox', '8-4']     4.2         35000   
4                                    ['Team Cherry']     4.4          3000   

   Number of Reviews                                             Genres  \
0              39000                               ['Adventure', 'RPG']   
1              29000           ['Adventure', 'Brawler', 'Indie', 'RPG']   
2              43000                               ['Adventure', 'RPG']   
3              35000  ['Adventure', 'Indie', 'RPG', 'Turn Based Stra...   
4               3000                 ['Adventure', 'Indie', 'Platform']   

                                             Summary  \
0  Elden Ring is a fantasy, action and open world...   
1  A rogue-lite hack and slash dungeon crawler in...   
2  The Legend of Zelda: Breath of the Wild is the...   
3  A small child falls into the Underground, wher...   
4  A 2D metroidvania with an emphasis on close co...   

                                             Reviews  Plays  Playing  \
0  ["The first playthrough of elden ring is one o...  17000    38000   
1  ['convinced this is a roguelike for people who...  21000    32000   
2  ['This game is the game (that is not CS:GO) th...  30000    25000   
3  ['soundtrack is tied for #1 with nier automata...  28000      679   
4  ["this games worldbuilding is incredible, with...  21000    24000   

   Backlogs  Wishlist  
0     46000     48000  
1     63000     36000  
2      5000     26000  
3     49000     18000  
4     83000     23000

Información del contenido de las columnas

Title - Title of the Game.

Release Date - Date of release of the game's first version

Team - Game developer team

Rating - Average rating

Times Listed - Number of users who listed this game

Number of Reviews - Number of reviews received from the users

Genres - All genres pertaining to a specified game

Summary - Summary provided by the team

Reviews - User reviews

Plays - Number of users that have played the game before

Playing - Number of current users who are playing the game

Backlogs - Number of users who have access but haven't started with the game yet

Wishlist- Number of users who wish to play the game

In [10]:
games.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1512 entries, 0 to 1511
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Title              1512 non-null   object 
 1   Release Date       1512 non-null   object 
 2   Team               1511 non-null   object 
 3   Rating             1499 non-null   float64
 4   Times Listed       1512 non-null   int64  
 5   Number of Reviews  1512 non-null   int64  
 6   Genres             1512 non-null   object 
 7   Summary            1511 non-null   object 
 8   Reviews            1512 non-null   object 
 9   Plays              1512 non-null   int64  
 10  Playing            1512 non-null   int64  
 11  Backlogs           1512 non-null   int64  
 12  Wishlist           1512 non-null   int64  
dtypes: float64(1), int64(6), object(6)
memory usage: 165.4+ KB


Tenemos pocas columnas numéricas. Vamos a transformar a columnas numéricas Times Listed, Number of Reviews, Plays, Playing, Backlogs,	Wishlist	

In [11]:
""" Este es el código utilizado para la modificación de columnas numéricas, eliminación de la k y sustitución por las unidades
correspondientes. Posteriormente al cambio guardamos el archivo, por lo que dejamos el código comentado porque no es necesario volver
a ejecutarlo"""

# def limpieza(col):
#     if "K" in col:
#         return (col.replace("K","000").replace(".", ""))
#     else:
#         return col

# games["Times Listed"] = games["Times Listed"].apply(limpieza)
# games["Number of Reviews"] = games["Number of Reviews"].apply(limpieza)
# games["Plays"] = games["Plays"].apply(limpieza)
# games["Playing"] = games["Playing"].apply(limpieza)
# games["Backlogs"] = games["Backlogs"].apply(limpieza)
# games["Wishlist"] = games["Wishlist"].apply(limpieza)

# games = games.astype({"Times Listed": int, "Number of Reviews": int, "Plays": int, "Playing": int, "Backlogs": int, "Wishlist":int})
    

' Este es el código utilizado para la modificación de columnas numéricas, eliminación de la k y sustitución por las unidades\ncorrespondientes. Posteriormente al cambio guardamos el archivo, por lo que dejamos el código comentado porque no es necesario volver\na ejecutarlo'

2. Extraer los siguentes estadísiticos del conjunto de datos para varias de las variables disponibles:
- Media
- Desviación con respecto de la media
- Varianza
- Desviación típica
- Moda
- Mediana
- Media ponderada
- Robustez
- Coeficiente de variación


In [ ]:
def descripcion(col):

    print(f"La media de la columna {col} es", games[col].mean())
    print(f"La desviación con respecto a la media de la columna {col} es: ", np.abs(games[col] - games[col].mean()))
    print(f"La varianza de la columna {col} es", games[col].var())
    print(f"La desviación estándar de la columna {col} es", games[col].std())
    print(f"La moda de la columna {col} es", games[col].mode()[0])
    print(f"La mediana de la columna {col} es", games[col].median())
    print(f"El coeficiente de variación de la columna {col} es", (games[col].std() / games[col].mean()))

In [ ]:
descripcion("Rating")

La media de la columna Rating es 3.7193462308205465
La desviación con respecto a la media de la columna Rating es:  0       0.780654
1       0.580654
2       0.680654
3       0.480654
4       0.680654
          ...   
1507    0.519346
1508    0.819346
1509    0.019346
1510    0.380654
1511    0.919346
Name: Rating, Length: 1512, dtype: float64
La varianza de la columna Rating es 0.2836708673606179
La desviación estándar de la columna Rating es 0.5326076110614811
La moda de la columna Rating es 4.1
La mediana de la columna Rating es 3.8
El coeficiente de variación de la columna Rating es 0.14319925546269444


In [ ]:
descripcion("Number of Reviews")

La media de la columna Number of Reviews es 4352.79298941799
La desviación con respecto a la media de la columna Number of Reviews es:  0       34647.207011
1       24647.207011
2       38647.207011
3       30647.207011
4        1352.792989
            ...     
1507     4258.792989
1508     4088.792989
1509     4142.792989
1510     4187.792989
1511     4168.792989
Name: Number of Reviews, Length: 1512, dtype: float64
La varianza de la columna Number of Reviews es 61362585.79629813
La desviación estándar de la columna Number of Reviews es 7833.427461609518
La moda de la columna Number of Reviews es 11000
La mediana de la columna Number of Reviews es 551.0
El coeficiente de variación de la columna Number of Reviews es 1.7996324384488873


In [ ]:
descripcion("Plays")

La media de la columna Plays es 30211.3167989418
La desviación con respecto a la media de la columna Plays es:  0       13211.316799
1        9211.316799
2         211.316799
3        2211.316799
4        9211.316799
            ...     
1507    29448.316799
1508    15211.316799
1509    19211.316799
1510    29942.316799
1511    13211.316799
Name: Plays, Length: 1512, dtype: float64
La varianza de la columna Plays es 691534493.7321335
La desviación estándar de la columna Plays es 26297.043440891477
La moda de la columna Plays es 12000
La mediana de la columna Plays es 21000.0
El coeficiente de variación de la columna Plays es 0.8704368503994693


In [ ]:
descripcion("Playing")

La media de la columna Playing es 1035.2367724867725
La desviación con respecto a la media de la columna Playing es:  0       36964.763228
1       30964.763228
2       23964.763228
3         356.236772
4       22964.763228
            ...     
1507     1030.236772
1508      986.236772
1509      990.236772
1510      956.236772
1511     1024.236772
Name: Playing, Length: 1512, dtype: float64
La varianza de la columna Playing es 15813763.439599883
La desviación estándar de la columna Playing es 3976.652290507668
La moda de la columna Playing es 11000
La mediana de la columna Playing es 112.5
El coeficiente de variación de la columna Playing es 3.841297369060061


In [ ]:
descripcion("Backlogs")

La media de la columna Backlogs es 11244.839285714286
La desviación con respecto a la media de la columna Backlogs es:  0       34755.160714
1       51755.160714
2        6244.839286
3       37755.160714
4       71755.160714
            ...     
1507    11021.839286
1508    10831.839286
1509    10757.839286
1510    10856.839286
1511    11005.839286
Name: Backlogs, Length: 1512, dtype: float64
La varianza de la columna Backlogs es 206351112.83517316
La desviación estándar de la columna Backlogs es 14364.926482066421
La moda de la columna Backlogs es 15000
La mediana de la columna Backlogs es 1000.0
El coeficiente de variación de la columna Backlogs es 1.277468367228331


In [ ]:
descripcion("Wishlist")

La media de la columna Wishlist es 4866.25462962963
La desviación con respecto a la media de la columna Wishlist es:  0       43133.74537
1       31133.74537
2       21133.74537
3       13133.74537
4       18133.74537
           ...     
1507     4799.25463
1508     4759.25463
1509     4660.25463
1510     4264.25463
1511     4793.25463
Name: Wishlist, Length: 1512, dtype: float64
La varianza de la columna Wishlist es 77247433.51354572
La desviación estándar de la columna Wishlist es 8789.051912097557
La moda de la columna Wishlist es 13000
La mediana de la columna Wishlist es 496.0
El coeficiente de variación de la columna Wishlist es 1.8061224866004373


In [ ]:
"""Media ponderada de Number of Reviews. Balancear la media considerando asignar mayor peso a los juegos que tienen menos 
reviews desde el planteamiento de que unos pocos juegos con muchas reviews distorsionan el resultado.
Decidimos hacer esto, aunque sabemos que va a arrojar un resultado representativo de la muestra ya que no consideramos que 
en nuestro dataset haya ninguna variable de la que podamos extraer una media ponderada coherente o justificada.""" 

def definir_pesos(col):
    if col < 100:
        return 0.40
    elif col < 500:
        return 0.30
    elif col <= 1000:
        return 0.20
    else:
        return 0.10

pesos = games["Number of Reviews"].apply(definir_pesos)
number_reviews_mp = np.average(games["Number of Reviews"], weights = pesos)
print(f"La media ponderada de Number of Reviews es {number_reviews_mp}")

La media ponderada de Number of Reviews es 2128.0612602100355


    Dado que para esta columna nuestra media aritmética es 4352.79 y la mediana 551.0, aunque la ponderación haya sido un poco "ficticia", consideramos que la media ponderada obtenida consigue el objetivo de acercarse algo más a la realidad de los datos. 

3. Interpretación de los resultados.
- Ahora interpreta los resultados obtenidos de los diferentes estadísticos estudiados haciendo hincapié en aspectos interesantes de tus datos y explica lo mejor posible la naturaleza de los mismos.

Estas son algunas de nuestras observaciones tras el estudio y trabajo con el dataset:

- Exceptuando la columna de ratings el resto de nuestras columnas numéricas son muy asimétricas (como ejemplifican las diferencias entre sus medias y medianas) y tienen bastantes outliers.
- Todas estas columnas numéricas están a su vez bastante relacionadas entre sí, en el sentido, de que comparten valores grandes o valores pequeños según el caso.
- Hemos visto que además dos de las columnas "Times Listed" y "Number of Reviews", aunque tienen diferente nombres, los datos son iguales. 
- Se observa que el grueso del peso en los estadísticos lo llevan en realidad unos pocos valores, lo que ejemplifica el tipo de industria de que se trata, con unos pocos juegos con mucho peso (como sucede en otras, música, libros...). 
- Creemos que sería interesante explorar el dataset por intervalos porque quizás disminuiría este efecto y arrojaría información interesante y más amplia sobre la industria del videojuego. 
- Resulta interesante la media del rating, 3.7, con una moda de 4.1, que nos lleva a pensar si existe tendencia a dejar más opiniones cuando estás son positivas, contrariamente a lo que podemos pensar en otros ámbitos. 

4. Guardamos el dataset modificado para conservar los cambios

In [ ]:
games.to_csv("data/games.csv")